client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx

```diff
@@ -25,27 +30,44 @@ const useIsRequestingSelectedSite = ( siteSlugOrId: string | number, site: SiteD
 };
 
 interface Props {
-	flowName: string;
+	flow: Flow;
 	stepSlug: string;
 	// If true, the tracking event will not be recorded
 	skipTracking: boolean;
-	flowVariantSlug?: string;
 }
 
 /**
  * Hook to track the step route in the declarative flow.
  */
-export const useStepRouteTracking = ( {
-	flowName,
-	stepSlug,
-	skipTracking,
-	flowVariantSlug,
-}: Props ) => {
+export const useStepRouteTracking = ( { flow, stepSlug, skipTracking }: Props ) => {
 	const intent = useIntent();
 	const design = useSelectedDesign();
 	const { site, siteSlugOrId } = useSiteData();
 	const isRequestingSelectedSite = useIsRequestingSelectedSite( siteSlugOrId, site );
 	const hasRequestedSelectedSite = siteSlugOrId ? !! site && ! isRequestingSelectedSite : true;
+	const signupStepStartProps = useSnakeCasedKeys( {
+		input: flow.useTracksEventProps?.()?.[ STEPPER_TRACKS_EVENT_SIGNUP_STEP_START ],
+	} );
+
+	const recordStepStart = useCallback(
+		( step: string ) => {
+			recordTracksEvent( STEPPER_TRACKS_EVENT_SIGNUP_STEP_START, {
+				flow: flow.name,
+				step: kebabCase( step ),
+				device: resolveDeviceTypeByViewPort(),
+				intent,
+				is_in_hosting_flow: isAnyHostingFlow( flow.name ),
+				...( design && { assembler_source: getAssemblerSource( design ) } ),
+				...( flow.variantSlug && { flow_variant: flow.variantSlug } ),
+				...signupStepStartProps,
+			} );
+		},
+
+		// We leave out intent and design from the dependency list, due to the ONBOARD_STORE being reset in the exit flow.
+		// The store reset causes these values to become empty, and may trigger this event again.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ flow.name, flow.variantSlug, signupStepStartProps ]
```

Comment:
Is there a risk of `signupStepStartProps` causing too many run of the effect?

Reply:
Good point. Yes, there is that risk. So the way we have things right now, and once the [steps param is cleaned up](https://github.com/Automattic/wp-calypso/issues/93808#issue-2480245630), this value should be memoized as long as the flow memoizes the initial input. Unfortunately the latter is required - and unless we can come up with some other way, we should make it clear in a doc somewhere (to always memo the return).

I'll give it more thought, but that's the same risk with wherever these extended props are used (they are an object).